eth/stagedsync/stage_bodies.go

```diff
@@ -252,15 +353,35 @@ Loop:
 			log.Trace("bodyLoop woken up by the incoming request")
 		}
 		d6 += time.Since(start)
+
+		return false, nil
 	}
-	if err := s.Update(tx, bodyProgress); err != nil {
-		return err
+
+	// kick off the loop and check for any reason to stop and break early
+	for !stopped {
+		shouldBreak, err := loopBody()
+		if err != nil {
+			return err
+		}
+		if shouldBreak {
+			break
+		}
 	}
+
+	// remove the temporary bucket for bodies stage
 	if !useExternalTx {
-		if err := tx.Commit(); err != nil {
+		bucketTx, err := cfg.db.BeginRw(context.Background())
+		if err != nil {
 			return err
 		}
+		defer bucketTx.Rollback()
+
+		bucketTx.ClearBucket("BodiesStage")
+		bucketTx.Commit()
```

Comment:
need check error returned here?

Reply:
Hopefully the bucket won't be too large, it depends a lot on the order in which the bodies are sent to Erigon though. The number of requests is somewhat lowered now that we are checking timeouts though so it shouldn't chase off too quickly up to 32k requests.  If it becomes an issue we could always check the bucket size before issuing a request and loop until some data has come in to process.